core/src/avm1/globals/bitmap_data.rs

```diff
@@ -0,0 +1,1120 @@
+        return Err(Error::ConstructorFailure);
+    }
+
+    let transparency = args
+        .get(2)
+        .unwrap_or(&Value::Bool(true))
+        .as_bool(activation.current_swf_version());
+
+    let fill_color = args
+        .get(3)
+        // can't write this in hex
+        // 0xFFFFFFFF as f64;
+        .unwrap_or(&Value::Number(4294967295_f64))
+        .coerce_to_i32(activation)?;
+
+    let bitmap_data = this.as_bitmap_data_object().unwrap();
+    bitmap_data.init_pixels(
+        activation.context.gc_context,
+        width as u32,
+        height as u32,
+        fill_color,
+    );
+    bitmap_data.set_transparency(activation.context.gc_context, transparency);
+
+    Ok(Value::Undefined)
+}
+
+pub fn get_height<'gc>(
+    _activation: &mut Activation<'_, 'gc, '_>,
+    this: Object<'gc>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    let bitmap_data = this.as_bitmap_data_object().unwrap();
+
+    if bitmap_data.get_disposed() {
+        return Ok((-1).into());
+    }
+
+    Ok(this.as_bitmap_data_object().unwrap().get_height().into())
+}
+
+pub fn get_width<'gc>(
+    _activation: &mut Activation<'_, 'gc, '_>,
+    this: Object<'gc>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    let bitmap_data = this.as_bitmap_data_object().unwrap();
+
+    if bitmap_data.get_disposed() {
+        return Ok((-1).into());
+    }
+
+    Ok(this.as_bitmap_data_object().unwrap().get_width().into())
+}
+
+pub fn get_transparent<'gc>(
+    _activation: &mut Activation<'_, 'gc, '_>,
+    this: Object<'gc>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    let bitmap_data = this.as_bitmap_data_object().unwrap();
+
+    if bitmap_data.get_disposed() {
+        return Ok((-1).into());
+    }
+
+    Ok(this
+        .as_bitmap_data_object()
+        .unwrap()
+        .get_transparency()
+        .into())
+}
+
+pub fn get_rectangle<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    this: Object<'gc>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    let bitmap_data = this.as_bitmap_data_object().unwrap();
+
+    if bitmap_data.get_disposed() {
+        return Ok((-1).into());
+    }
+
+    let proto = activation.context.system_prototypes.rectangle_constructor;
+    let rect = proto.construct(
+        activation,
+        &[
+            0.into(),
+            0.into(),
+            bitmap_data.get_width().into(),
+            bitmap_data.get_height().into(),
+        ],
+    )?;
+    Ok(rect.into())
+}
+
+pub fn get_pixel<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    this: Object<'gc>,
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    let bitmap_data = this.as_bitmap_data_object().unwrap();
+
+    if bitmap_data.get_disposed() {
+        return Ok((-1).into());
+    }
+
+    let x = args.get(0).and_then(|x| x.coerce_to_i32(activation).ok());
+    let y = args.get(1).and_then(|x| x.coerce_to_i32(activation).ok());
```

Comment:
`ok()` swallows an error, instead there should be `?` somewhere, here's a pathological example:
```
var o = {valueOf: function() {
	throw "BAD";
	return 3;
}};
trace(bData.getPixel(o, o));
```